### PR TITLE
ENG-3461: migrate admin-ui to RouterLink and remove legacyBehavior

### PR DESCRIPTION
**Pull request by gilluminate**
Ticket [ENG-3461]

### Description Of Changes

Next.js 16 dropped `passHref` and `legacyBehavior` from the public `<Link>` API, and Next.js 15 emits a runtime deprecation warning for `legacyBehavior`. The Admin UI still relied on that pattern in **20 places** to wrap a `Typography.Link` or `Button` inside a `next/link`.

This PR introduces a shared `RouterLink` component (`clients/admin-ui/src/features/common/nav/RouterLink.tsx`) that replaces every `<Link passHref legacyBehavior>` call site in admin-ui. It detects whether its only child is an antd `Button` and either:

- wraps the button in `next/link` so Next renders its own `<a>` around the button (preserves prefetch + modifier-click handling), or
- renders a `Typography.Link`-styled anchor that intercepts plain left-clicks and calls `router.push` for SPA navigation, while letting modifier / middle / right clicks fall through to the browser for new-tab / copy-link behaviour.

Also forwards `target` and `rel` to both paths so callers can still open a link in a new tab (bypassing SPA interception when `target="_blank"`).

The result is a single anchor per link (no more nested `<a>` hack), no deprecation warnings, and consistent SPA navigation across the app. Unit tests cover both branches of the component.

After this PR, `git grep 'passHref\|legacyBehavior'` in `clients/admin-ui` returns zero hits. Privacy Center, fides-js, and fidesui already had zero deprecated Link usages and did not need changes.

### Code Changes

* Added `clients/admin-ui/src/features/common/nav/RouterLink.tsx` + jest coverage in `RouterLink.test.tsx` (10 tests: Button-wrap path, text path, URL-object serialisation, modifier-click pass-through, `target="_blank"` pass-through, `target`/`rel` forwarding, custom `onClick` + `preventDefault`, fragment fall-through).
* Migrated **20** call sites off `<Link passHref legacyBehavior>`:
  * `features/access-policies/PolicyCard.tsx`
  * `features/common/table/cells/LinkCell.tsx`
  * `features/consent-settings/tcf/PublisherRestrictionsTable.tsx`
  * `features/data-discovery-and-detection/action-center/ConfidenceCard.tsx`
  * `features/data-discovery-and-detection/action-center/EmptyMonitorsResult.tsx`
  * `features/data-discovery-and-detection/action-center/MonitorResult.tsx`
  * `features/data-discovery-and-detection/action-center/fields/page.tsx`
  * `features/datamap/datamap-drawer/SystemInfo.tsx`
  * `features/integrations/IntegrationLinkedSystems.tsx`
  * `features/privacy-assessments/AssessmentSettingsModal.tsx` (uses `target="_blank"`)
  * `features/privacy-requests/dashboard/DuplicateRequestsButton.tsx`
  * `features/user-management/RolesForm.tsx`
  * `features/user-management/UserManagementTable.tsx`
  * `home/SystemCoverageCard.tsx` (text-mode link)
  * `pages/access-policies/index.tsx`
  * `pages/add-systems/multiple.tsx`
  * `pages/dataset/index.tsx`
  * `pages/privacy-assessments/[id].tsx`
  * `pages/settings/rbac/index.tsx`
  * `pages/settings/rbac/roles/new.tsx`

### Steps to Confirm

All of the below were verified manually against a local `npm run dev` admin-ui. After each, confirm the URL changes to the expected destination and no `validateDOMNesting` / nested-anchor / `legacyBehavior` / `passHref` warnings appear in the console.

1. Go to `/dataset` → click the **+ Add dataset** button (top right) → lands on `/dataset/new`.
2. Go to `/add-systems/multiple` → click the inline **Add a system** link in the description → lands on `/add-systems/manual`.
3. Go to `/data-discovery/action-center` → on any monitor card click **Review** → lands on `/data-discovery/action-center/{website|datastore|infrastructure}/{monitorId}`.
4. Go to `/user-management` → click **Add new user** (top right) → lands on `/user-management/new`.
5. Go to `/integrations` → on any row, click the integration name in the first column (should be a single blue link, ellipsis-truncated, and the row itself should NOT also navigate) → lands on `/integrations/{id}`.
6. On an integration detail page that has a linked system (e.g. the BigQuery Connection seed data), open the **Linked system** tab → click the system name in the list → lands on `/systems/configure/{systemKey}`.
7. Go to `/reporting/datamap` → click a system cell in the table to open the drawer → click **View more** in the drawer header → lands on `/systems/configure/{systemKey}`.
8. Go to `/settings/consent` on a TCF config with restrictions → click any **Edit** button in the restrictions table → lands on `/settings/consent/{configId}/{restrictionId}`.
9. Go to `/` (home dashboard) → on the **System Coverage** card, click **Connect more systems** in the card header (text link, Typography.Link style) → lands on `/add-systems/manual`.

Modifier-click check (optional but worth it): on any of the above links, Cmd/Ctrl-click should open the target in a new tab (not SPA-navigate in place).

### Pre-Merge Checklist

* [x] Issue requirements met
* [x] All CI pipelines succeeded
* [x] `CHANGELOG.md` updated
  * [ ] Add a https://github.com/ethyca/fides/labels/db-migration label to the entry if your change includes a DB migration
  * [ ] Add a https://github.com/ethyca/fides/labels/high-risk label to the entry if your change includes a high-risk change (i.e. potential for performance impact or unexpected regression) that should be flagged
  * [ ] Updates unreleased work already in Changelog, no new entry necessary
* UX feedback:
  * [ ] All UX related changes have been reviewed by a designer
  * [x] No UX review needed
* Followup issues:
  * [ ] Followup issues created
  * [x] No followup issues
* Database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!
  * [x] No migrations
* Documentation:
  * [ ] Documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] Documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] If there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
  * [x] No documentation updates required

[ENG-3461]: https://ethyca.atlassian.net/browse/ENG-3461?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: clients/admin-ui/src/features/common/nav/RouterLink.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import { Button, Typography } from "fidesui";
+import { formatUrl } from "next/dist/shared/lib/router/utils/format-url";
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/src/features/common/nav/RouterLink.tsx]: Good catch! Dropped the `next/dist` import and replaced it with a small inline `formatHref` helper that serialises `pathname` + `query`/`search` + `hash` via `URLSearchParams`. Added 12 unit tests covering single/multi/array query values, null/undefined skipping, URL encoding, hash prefixing, explicit `search` precedence, and string queries. No more private-API dependency.

File: clients/admin-ui/src/features/common/nav/RouterLink.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  "href" | "onClick"
+>;
+
+export interface RouterLinkProps extends TypographyLinkProps {
+  href: Href;
+  children: ReactNode;
+  onClick?: (e: MouseEvent<HTMLElement>) => void;
+  /** Forwarded to next/link when the child is a Button. */
+  replace?: NextLinkProps["replace"];
+  /** Forwarded to next/link when the child is a Button. */
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/src/features/common/nav/RouterLink.tsx]: Wired `replace`, `scroll`, and `prefetch` through to text mode rather than dropping them. `replace` switches to `router.replace`, `scroll` is forwarded via the options arg, and `prefetch` mirrors `next/link` semantics: skip on `false`, hover-only on `null`, mount-prefetch otherwise (production only). Added 5 tests covering each branch.

File: clients/admin-ui/src/features/common/nav/RouterLink.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  prefetch?: NextLinkProps["prefetch"];
+}
+
+const isAntButtonChild = (children: ReactNode): boolean => {
+  const arr = Children.toArray(children);
+  if (arr.length !== 1) {
+    return false;
+  }
+  const [only] = arr;
+  return isValidElement(only) && only.type === Button;
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/src/features/common/nav/RouterLink.tsx:69]: Yep, intentional. The structural detection is documented in the JSDoc, and any HOC/memo-wrapped Button just falls through to text mode (still functional, just not the wrapped-button variant). If we hit a real call site that needs it, we can add a `button` escape-hatch prop.

File: clients/admin-ui/src/features/data-discovery-and-detection/action-center/MonitorResult.tsx [hunk 4]
```diff
@@ -17,6 +17,7 @@ import palette from "fidesui/src/palette/palette.module.scss";
 import NextLink from "next/link";
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/src/features/data-discovery-and-detection/action-center/MonitorResult.tsx:17]: Correct, intentional. The other `<NextLink>` usage further down doesn't fit the legacyBehavior wrap pattern and stays on plain `next/link`.

File: clients/admin-ui/cypress/e2e/action-center/aggregate-results.cy.ts [hunk 5]
```diff
@@ -81,16 +81,20 @@ describe("Action center", () => {
...
-      );
+      cy.getByTestId(`review-button-${webMonitorKey}`)
+        .parent("a")
+        .should(
+          "have.attr",
+          "href",
+          `${ACTION_CENTER_ROUTE}/${APIMonitorType.WEBSITE}/${webMonitorKey}`,
+        );
+      cy.getByTestId(`review-button-${integrationMonitorKey}`)
+        .parent("a")
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/cypress/e2e/action-center/aggregate-results.cy.ts]: Done! Switched both to `.closest("a")` so the test survives any intermediate wrapper changes.

**Review by rayharnett — COMMENTED** [review]
# Code Review: PR 7942

**Branch:** `gill/ENG-2987/link-legacy-behavior`  
**Status:** ✅ LGTM (Looks Good To Me)

## Summary of Changes
The PR introduces a new shared `RouterLink` component in the Admin UI to replace usage of `next/link` with `legacyBehavior` and `passHref`. This change aims to unify internal navigation and simplify how components like Ant Design's `Button` are wrapped for client-side routing.

## Key Improvements

### 1. Unified Navigation Logic
The new `RouterLink` handles both plain text (using `Typography.Link`) and component wrapping (specifically detecting single Ant Design `Button` children) to ensure correct HTML structure without nested anchors. This solves the issue where `Next.js` and custom components would produce invalid nested `<a>` tags.

### 2. Improved Developer Experience
Developers no longer need to remember to use `legacyBehavior` or `passHref` when wrapping custom components. The component intelligently detects its children to apply the correct rendering strategy.

### 3. Enhanced Functionality
- **URL Serialization**: Built-in support for `UrlObject` serialization via a robust `formatHref` utility that handles complex query objects, including arrays and null/undefined values.
- **Intelligent Prefetching**: Supports prefetching on mount (in production) and provides an optimized `onMouseEnter` prefetch strategy for non-production environments to improve perceived performance.
- **Native Browser Compatibility**: Correctly handles modifier keys (meta, ctrl, shift, alt) and middle clicks, allowing standard browser features like "Open in new tab" or "Copy link" to function as expected.

### 4. Comprehensive Testing
A robust test suite (`RouterLink.test.tsx`) covers:
- `formatHref` edge cases (query serialization, hash handling, encoding).
- `RouterLink` interaction patterns (Button wrapping, text mode, prefetching, and modifier key interception).

## Detailed Review Findings

### 🟢 Pass: Technical Correctness & Architecture
- **Structural Detection**: The use of `isAntButtonChild` to decide between wrapping with `NextLink` or using a styled anchor is an effective way to avoid the "nested anchor" issue in modern Next.js versions.
- **Event Handling**: The implementation of `onClick` within `RouterLink` correctly intercepts only standard left-clicks while allowing native browser navigation for modifier keys. This is crucial for accessibility and UX.
- **Refactoring Scope**: The replacement of `NextLink` across multiple features (`access-policies`, `consent-settings`, `data-discovery`, etc.) is thorough and consistent.

### 🟡 Minor Observations / Suggestions
- **Complexity in Detection**: The `isAntButtonChild` logic relies on a strict structural check (a single child that is exactly the `Button` component). While effective, this is "magic" behavior. It's recommended to ensure any future custom components intended for use here are either wrapped in a fragment or handled via an explicit prop if they don't meet this specific criteria.
- **Prefetching Granularity**: The `onMouseEnter` prefetching is a great optimization. As the application grows, we should monitor that large tables with many links do not cause excessive network noise during rapid mouse movement.

### 🔴 Security & Compliance (Zero Tolerance Check)
- **Sensitive Data**: Verified that no real customer data, internal URLs (`*.ethyca.com`), or private credentials were introduced in the code or test data.
- **Compliance**: The PR adheres to all project security guidelines regarding placeholder usage and generic error messages.

## Final Verdict
The refactor is well-architected, thoroughly tested, and provides a significant improvement to the codebase's maintainability and consistency by abstracting away the complexities of Next.js link behavior.

**Review by rayharnett — APPROVED** [review]
Approved, see prior review.
